libraries/standard/mqtt/integration-test/mqtt_system_test.c

```diff
@@ -836,15 +864,78 @@ void test_MQTT_ProcessLoop_KeepAlive( void )
 {
     uint32_t connectPacketTime = context.lastPacketTime;
     uint32_t elapsedTime = 0;
+
     TEST_ASSERT_EQUAL( 0, context.pingReqSendTimeMs );
 
-    /* Sleep until control packet needs to be sent. */
+    /*Sleep until control packet needs to be sent. */
     Clock_SleepMs( MQTT_KEEP_ALIVE_INTERVAL_SECONDS * 1000 );
     TEST_ASSERT_EQUAL( MQTTSuccess, MQTT_ProcessLoop( &context, MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
 
     TEST_ASSERT_NOT_EQUAL( 0, context.pingReqSendTimeMs );
     TEST_ASSERT_NOT_EQUAL( connectPacketTime, context.lastPacketTime );
-    /* Test that the ping was sent within 1.5 times the keep alive interval. */
+    /*Test that the ping was sent within 1.5 times the keep alive interval. */
     elapsedTime = context.lastPacketTime - connectPacketTime;
     TEST_ASSERT_LESS_OR_EQUAL( MQTT_KEEP_ALIVE_INTERVAL_SECONDS * 1500, elapsedTime );
 }
+
+void test_MQTT_Connect_Restore_Session( void )
+{
+    /* Graceless disconnect. Terminate TLS session and TCP connection. */
+    ( void ) Openssl_Disconnect( &networkContext );
+
+    /* Establish a new MQTT connection over TLS with the broker with the "clean session" flag set to 0
+     * to start a persistent session with the broker. */
+
+    /* Create the TLS+TCP connection with the broker. */
+    TEST_ASSERT_EQUAL( OPENSSL_SUCCESS, Openssl_Connect( &networkContext,
+                                                         &serverInfo,
+                                                         &opensslCredentials,
+                                                         TRANSPORT_SEND_RECV_TIMEOUT_MS,
+                                                         TRANSPORT_SEND_RECV_TIMEOUT_MS ) );
+    TEST_ASSERT_NOT_EQUAL( -1, networkContext.socketDescriptor );
+    TEST_ASSERT_NOT_NULL( networkContext.pSsl );
+
+    /* Establish a new MQTT connection for a persistent session with the broker. */
+    establishMqttSession( &context, false, &networkContext, false, &persistentSession );
+    TEST_ASSERT_FALSE( persistentSession );
+
+    /* Publish to a topic with Qos 2. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, publishToTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS2 ) );
+
+    /* Disconnect on receiving PUBREC so that we aren't able to complete the QoS 2 PUBLISH in the current connection. */
+    disconnectOnPacketType = MQTT_PACKET_TYPE_PUBREC;
+
+    TEST_ASSERT_EQUAL( MQTTSendFailed,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+    /*TEST_ASSERT_TRUE( receivedPubRec ); */
+    TEST_ASSERT_FALSE( receivedPubRel );
```

Comment:
Does this line need to be here? We receive a pubrel for an incoming publish, be we disconnect during the QoS 2 flow of an outgoing publish. Is it a protocol error for the client to receive both an incoming publish and pubrel before it receives a pubrec for the outgoing publish? If not, I don't think we need this assert.

Reply:
Yes, doesn't need to be here. It is an artifact from an older state of the code 😉